Add detailed man page documentation and minor fixes

davidallendj opened the pull request:
This PR adds man pages documentation as well as some minor fixes to the cobra CLI help strings. Each man page helps to further document the behavior of the CLI detailing more about how each flag works with information that is not included in the help strings (i.e. documentation from `--help`). 

Edit: This PR also adds a couple of minor fixes, such as `scan` not writing to the database properly.

To generate and view the docs: 

1. Install `scdoc` using package manager
2. Generate the docs
```bash
scdoc < man/magellan.1.sc > docs/magellan.1
scdoc < man/magellan-scan.1.sc > docs/magellan-scan.1
scdoc < man/magellan-list.1.sc > docs/magellan-list.1
scdoc < man/magellan-collect.1.sc > docs/magellan-collect.1
scdoc < man/magellan-crawl.1.sc > docs/magellan-crawl.1
scdoc < man/magellan-send.1.sc > docs/magellan-send.1
scdoc < man/magellan-update.1.sc > docs/magellan-update.1
scdoc < man/magellan-secrets.1.sc > docs/magellan-secrets.1
```
3. View the docs using `man`:
```bash
man docs/magellan.1
```

TODO:

- [x] Add rule to Makefile to generate docs
- [x] Fix syntax issues of *.sc docs

davidallendj committed:
docs: fix scan help string

davidallendj committed:
docs: added initial set up man pages

davidallendj committed:
docs: updated all man pages content and format

davidallendj committed:
docs: updated all man pages content and format again

davidallendj committed:
docs: fixed syntax formatting for scdoc

synackd commented:
Let me know if you would like help with the Makefile. I think the one for the ochami CLI is pretty tuned.

davidallendj commented:
@synackd I had already started copying from the ochami Makefile, but if you want to take over that would be appreciated.

davidallendj committed:
refactor: minor changes and fixes

davidallendj committed:
refactor: added String function to RemoteAsset

synackd committed:
chore(man): remove trailing spaces

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

davidallendj committed:
docs: correct variable for magellen-collect man page

davidallendj committed:
refactor: changed interface{} to any

synackd committed:
docs(man): fix scdoc compilation errors

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

synackd committed:
docs(man): add .gitignore

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

synackd committed:
docs(man): fix formatting

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

synackd committed:
build(Makefile): refactor and add man page targets

The config.env file was removed in favor of setting variables on the
command line e.g:

make prefix=/my/install/path install

Recipes also have been reordered alphabetically based on the target.

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

davidallendj commented on man/magellan-secrets.1.sc:
```diff
@@ -1,4 +1,4 @@
-MAGELLAN-SECRETS "OpenCHAMI" "Manual Page for magellan-secrets"
+MAGELLAN-SECRETS(1) "OpenCHAMI" "Manual Page for magellan-secrets"
```
Ahh so is that what was causing the compilation error?

synackd replied:
Looks like it. `scdoc` seems to require the section number.

synackd commented on cmd/update.go:
```diff
@@ -26,8 +26,7 @@ var updateCmd = &cobra.Command{
   magellan update 172.16.0.108:443 -i -u $bmc_username -p $bmc_password \
-    --firmware-url http://172.16.0.200:8005/firmware/bios/image.RBU \
-    --component BIOS
+    --firmware-uri http://172.16.0.200:8005/firmware/bios/image.RBU \
```
Since `--component BIOS` was removed, we can probably removed the trailing backslash from the `firmware-uri` line.

synackd commented on man/magellan-collect.1.sc:
```diff
@@ -0,0 +1,157 @@
+	made to BMC nodes. When this flag is not provided, the default system
+	certificates are used instead.
+
+	**DEPRECATED**
```
Is there an alternative? We might want to add a deprecation warning message here, and if this option will be removed in a future version.

davidallendj replied:
Nope. It's going to be removed from `collect` since it no longer makes a request so I'll add a message for it. That functionality was moved to `send` which has the flag too.

davidallendj replied:
Added message in https://github.com/OpenCHAMI/magellan/pull/124/commits/8a96a6d253aab403d89e4f28664b6827aea4ac9b.

davidallendj commented on man/magellan.1.sc:
I need to come back and add content to this file. I completely forgot to add stuff here.

synackd replied:
Sounds good. My placeholders were just so it wasn't empty, feel free to remove them.

synackd replied:
Do you plan on addressing these in this PR or in a subsequent one?

synackd commented on man/magellan-scan.1.sc:
```diff
@@ -0,0 +1,145 @@
+*--insecure*
+	Skip TLS certificate verification with the BMC when performing probing
+	requests. After finding remote assets on a network, a subsequent request is
+	made to determine which assets run a Redfish service. (default true)
```
Does `default true` mean that this flag is set by default (which would make passing it redundant) or that the connection is secure by default? I presume you meant the latter, and I think it would be good to clarify that here.

davidallendj replied:
I think it means the former looking at `cmd/scan.go`. I think the intent there was to skip the TLS check by default for scans but still allow `--insecure=false` as an option.

synackd replied:
In that case, I think it would help clarify if we specified that and included an explicit reference to `--insecure=false` here.

davidallendj replied:
After looking a bit closer at this, I'm now thinking the flag doesn't make sense at all. Whenever `--insecure=false`, we need to supply a cert with the request to the BMC, but there's no mechanism for doing so for `scan`. As it stands now, it may just be better to either remove the flag entirely or add a `--cacert` flag like `send`.

synackd replied:
The cases as I see are:

- the system bundle has a CA cert that trusts the BMC certs, use it with Magellan
- there is a separate CA cert for the BMCs, use it with Magellan
- don't use certs for the BMCs with Magellan

I think these cases are accommodated with both the `--insecure` and `--cacert` flags, but when TLS is enabled by default. If TLS is _disabled_ by default, then `--insecure` would be replaced with `--secure` for these cases. However, with security in mind, it's probably more prudent to have TLS be _enabled_ by default (secure by default), disabling it with the `--insecure` flag when not using certificates IMO.

There are probably a lot of folks out there (us included) who don't use certs, but it's probably better to be secure by default.

davidallendj replied:
I changed the default value and added more to the description in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b63eaf84939db59a3bb18bedebb91f08fd3e794.

synackd commented on man/magellan-send.1.sc:
```diff
@@ -0,0 +1,58 @@
+*-d, --data* -F _format_ (_node_object_,... | @_path_)
+	Specify node data objects to send to specified host. Objects can be loaded
+	from files using the '@' symbol followed by the path to the file. The input
+	format for (json, yaml)he objects can be specified to be either JSON or YAML by setting
```
I think the `t` in `the` accidentally got replaced by `(json, yaml)` here.

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/96f15b82faf0a86cd03992133f9374cfcc0935df.

synackd commented on man/magellan-send.1.sc:
````diff
@@ -0,0 +1,58 @@
+MAGELLAN-SEND(1) "OpenCHAMI" "Manual Page for magellan-send"
+
+# NAME
+
+magellan-send - Send node data from collect output to remote host
+
+# SYNOPSIS
+
+magellan send [OPTIONS] _host_
+
+# EXAMPLES
+
+// minimal working example++
+magellan send -d @inventory.json https://smd.openchami.cluster
+
+// send data from multiple files (must specify -f/--format if not JSON)++
+magellan send -d @cluster-1.json -d @cluster-2.json https://smd.openchami.cluster++
+magellan send -d '{...}' -d @cluster-1.json https://proxy.example.com
+
+// send data to remote host by piping output of collect directly++
+magellan collect -v -F yaml | magellan send -d @inventory.yaml -F yaml https://smd.openchami.cluster
+
+# FLAGS
+
+*--cacert* string
+	Set the path to CA cert file (defaults to system CAs when blank)
+
+*-d, --data* -F _format_ (_node_object_,... | @_path_)
+	Specify node data objects to send to specified host. Objects can be loaded
+	from files using the '@' symbol followed by the path to the file. The input
+	format for (json, yaml)he objects can be specified to be either JSON or YAML by setting
+	the *--format* flag.
+
+	An example of a node data object would look like the following using the
+	JSON format:
+
+	```
+
+	```
````
Missing example?

davidallendj replied:
Yeah, I think I was trying to get `collect` output but ended up running into a bug. Let me fix that.

davidallendj replied:
Added in https://github.com/OpenCHAMI/magellan/pull/124/commits/96f15b82faf0a86cd03992133f9374cfcc0935df.

synackd commented on man/magellan-update.1.sc:
```diff
@@ -0,0 +1,54 @@
+*--scheme* _scheme_
+	Specify the transfer protocol scheme to use for the request to the remote
+	host. Values are case-insensitive and converted to use upper-case letters.
+	Additionally, the default value for _scheme_ is HTTPS.
```
We might want to replace `HTTPS` here with `*https*` (bolded) if that is the literal value to be passed. I recall seeing it lower case elsewhere.

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/96f15b82faf0a86cd03992133f9374cfcc0935df.

synackd commented on pkg/scan.go:
```diff
@@ -61,10 +70,6 @@ func ScanForAssets(params *ScanParams) []RemoteAsset {
 		chanHosts = make(chan []string, params.Concurrency+1)
 	)
 
-	if params.Verbose {
-		log.Info().Any("args", params).Msg("starting scan...")
-	}
-
```
Should we put
```
log.Debug().Any("args", params).Msg("starting scan...")
```
here instead?

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/96f15b82faf0a86cd03992133f9374cfcc0935df.

davidallendj replied:
I remember why I took that out now. It prints all of the hosts+ports to test and is redundant when `--log-level=trace`. I changed it to print all the args except the hosts with `--log-level=debug` and with the hosts with `--log-level=trace` in https://github.com/OpenCHAMI/magellan/pull/124/commits/96795e68470ab2aa49e85382c17923e6979dd342.

synackd committed:
docs(man): fix formatting (part 2)

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

synackd reviewed (commented):
Initial manual pages look good! Just caught a few things to be addressed. I also refactored the Makefile and included a target to build the man pages.

```
make man
make clean-man
```

You can also build individual ones:

```
make man/magellan.1
```

davidallendj committed:
refactor: address comments in pull request

davidallendj committed:
refactor: changed how scan args are printed

davidallendj committed:
refactor: removed unused code in cmd/scan

davidallendj committed:
refactor: corrected scan params host field

synackd commented on man/magellan-scan.1.sc:
```diff
@@ -73,7 +73,10 @@ magellan scan --subnet 10.0.0.0 --subnet 172.16.0.0 --subnet-mask 255.255.0.0 --
+	made to determine which assets run a Redfish service. The default value for
+	this flag is *true*.
+
+	When *--insecure* is set to *false*, 
```
Incomplete description?

davidallendj replied:
Whoops! I was in the middle of writing this and got distracted. 🫠

davidallendj replied:
Completed description in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b63eaf84939db59a3bb18bedebb91f08fd3e794.

synackd replied:
Currently writing OpenCHAMI documentation myself, I completely understand. 😉

davidallendj committed:
docs: added deprecation warning message

davidallendj committed:
docs: added complete description to --insecure flag

davidallendj committed:
fix: issue with scan not writing to cache db

davidallendj committed:
fix: issue with list not printing output

davidallendj committed:
refactor: changed verbose to use zerolog instead

davidallendj committed:
docs: updated the example config file

davidallendj committed:
docs: updated example in secrets man page

davidallendj committed:
docs: updated main man page

synackd commented on man/magellan.1.sc:
```diff
@@ -79,11 +79,102 @@ The *magellan* command accepts
+For Redfish specifications and other documents, visit https://www.dmtf.org/standards/redfish.
 
 ## CONFIGURATION
 
-Nothing here yet.
+The *magellan* CLI configuration is handled by passing the *--config* in a single 
+YAML file. An example file can be found at in the root repository named
+*example.config.yaml*.
```
At some point, it would be useful to have a section 5 man page on the Magellan config file detailing format and available options.

davidallendj replied:
Sure, but for now I think the `example.config.yaml` and the flag description should be sufficient enough considering that all of the config properties have a corresponding flag.

synackd replied:
Agreed, just putting it out there to be complete.

synackd commented on man/magellan.1.sc:
```diff
@@ -79,11 +79,102 @@ The *magellan* command accepts
+The *magellan* CLI configuration is handled by passing the *--config* in a single 
+YAML file. An example file can be found at in the root repository named
+*example.config.yaml*.
+
+**
```
Extra empty bold specifier?

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b9208c45da69af93ce25d5409610b462a4c239d.

synackd commented on man/magellan.1.sc:
```diff
@@ -79,11 +79,102 @@ The *magellan* command accepts
+3. Collect Workflow:	scan -> collect -> send -> collect -> send -> *
+4. Crawl Workflow:		crawl -> send -> crawl -> send -> *
+
+### Simple Workflow
```
```
Error at 92:4: Only headings up to two levels deep are permitted
```

Could we move:
- `## GETTING STARTED` -> `# GETTING STARTED`
- `### * Workflow` -> `## * Workflow`

?

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b9208c45da69af93ce25d5409610b462a4c239d.

synackd commented on man/magellan.1.sc:
```diff
@@ -79,11 +79,102 @@ The *magellan* command accepts
+For more information about Redfish, visit https://https://redfish.dmtf.org/.
+For Redfish specifications and other documents, visit https://www.dmtf.org/standards/redfish.
 
 ## CONFIGURATION
```
To be consistent with `GETTING STARTED` above, can we move this to `# CONFIGURATION`?

davidallendj replied:
That doesn't quite seem appropriate to put these 2 lines there since this is more related to general information about Redfish. Could you explain your reasoning for moving this to that section? I don't think I quite understand.

synackd replied:
I think the GitHub diff context might be causing confusion (where did the line number indications go?). I'm referencing only the `## CONFIGURATION` line, not the ones above it.

davidallendj replied:
Oh I see. You're saying `## CONFIGURATION` -> `# CONFIGURATION`?

synackd replied:
Yeah. Just trying to see how we can solve the too-many-indentation-levels issue. My comment above for `GETTING STARTED` put it on `#` as well and I assume you want these at the same level.

davidallendj replied:
Yeah, that's right. I changed it in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b9208c45da69af93ce25d5409610b462a4c239d.

synackd commented on man/magellan.1.sc:
````diff
@@ -79,11 +79,102 @@ The *magellan* command accepts
+store set up and send the data to our remote host all in one step like in the
+"Simple Workflow".
+
+```
+# assume we already have a cache database and secrets store somewhere...
+magellan collect --secrets_file secrets.json | magellan send https://smd.example.com
+
+# alternatively, just update the file we send our output
+magellan collect --secrets-file secrets.json -o nodes.json
+```
````
We used `//` as the comments elsewhere. Can we do the same in this section to be consistent? That or change `//` elsewhere to be `#`.

davidallendj replied:
That was a mistake. I'll change it to `//`.

davidallendj replied:
Fixed in https://github.com/OpenCHAMI/magellan/pull/124/commits/3b9208c45da69af93ce25d5409610b462a4c239d and https://github.com/OpenCHAMI/magellan/pull/124/commits/54e9b7c2355ac625fcd9fe78ae504e3b5a85edba.

synackd commented on pkg/scan.go:
```diff
@@ -162,9 +162,7 @@ func ScanForAssets(params *ScanParams) []RemoteAsset {
-	if params.Verbose {
-		log.Info().Msg("scan complete")
-	}
+	log.Trace().Msg("scan complete")
```
I'm wondering if it makes more sense for this to be a debug-level message instead of a trace-level one.

davidallendj replied:
I made it debug-level at first, but I wasn't happy having the other message at the top of the function being trace-level and not print anything and then having this one being debug and showing by itself.

synackd reviewed (changes requested):
Largely looks good, there are just a few small things and an `scdoc` compilation error that needs addressing.

davidallendj committed:
docs: further addressed PR comments

davidallendj committed:
docs: fixed comment in main man page

synackd committed:
docs(man): magellan.1: fix newlines and REFERENCES case

Signed-off-by: Devon Bautista <17506592+synackd@users.noreply.github.com>

synackd reviewed (approved):
Fixed some newline issues I saw in **magellan**(1). Other than that, LGTM.
